lib/src/shared/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -979,10 +1018,13 @@ private void startReachabilityObserver() {
 
     // - (void) clearRepl
     private void clearRepl() {
-        if (c4repl != null) {
-            c4repl.free();
-            c4repl = null;
+        final C4Replicator repl;
+        synchronized (lock) {
+            if (c4Replicator == null) { return; }
+            repl = c4Replicator;
+            c4Replicator = null;
         }
+        repl.free();
```

Comment:
is there a reason why we keep `repl` instance outside lock?

Reply:
Not quite sure what your are asking....
We need the `repl` outside the lock, because we need to call `free` on it, right?
The idea is to hold locks for as little time as possible and to try not to hold multiple locks at one time.  Those are both Java idioms.  The first one prevents serializing processes that can run concurrently, the second prevents deadlocks.
In this particular case, there is no need to stop all further replicator processing, while the native object is freed.  Once the field is clear (null) the part that needs to be atomic is over, and the lock can be released.